### ISSUE TITLE
Feature Request: export .gif files

### ISSUE THREAD

**Issue by cahlander** [hsw107/Photoshop-Export-Layers-to-Files-Fast#51]
I've used this program to export PNG files of counties in Kansas (105 layers/counties) and the program is excellent!  One of my programs uses GIF files.  It uses GIF files since I can change the color of a layer by editing the text string for the GIF file.  Perhaps I can also change the color in a PDF file, but I don't know how.  If I exported the layers as GIF files they would work with my program.

Don't consider this request if it takes any effort.  I've read the .jsx file and see that many parameters are set up and than it looks like an internal PhotoShop routine is called.

Thanks.

**Comment by skjorn** [comment]
I'm glad you find it helpful. We could add the support for GIF eventually, sure, but it would take some time to handle all these parameters, as you say.

I'm not sure what "string" you're talking about, but the functionality you describe (changing colours) can be achieved using PNG-8 as well. It basically is the same thing -- it has a palette of colours and pixels reference the palette by indices. If you use a standard library (libpng, for example) in your program, you can change the colours in the palette and the whole picture will become different. Assuming you can modify your program. 

The only advantage the GIF has over the PNG is animation. And I'm afraid it wouldn't work well with this script. (Well, you could probably export multiple animations with one changed frame.)

**Comment by cahlander** [comment]
I'm going to take your advice and see if I can change the color of a PNG 
file.

I'll show you what I have.  There is an include file that has the base64 
encoding of the GIF files (map_medium.php) an example of an entry is:

/* 20 Aitkin */ 
'R0lGODlhuQDbAIAAAP8AAP///yH5BAEAAAEALAAAAAC5ANsAAAL/jI+py+0Po5y02ouz3rz7D4biSJbmiabqyrbuC8fyTNf2jef6zvf+DwwKh8Si8YhMKpfMpvMJjUqn1Kr1is1qt9yu9wsOi8fksvmMTqvX7Lb7DY/L5/S6/Y7P6/f8vv8PGCg4SFhoeIiYqLjI2Oj4CBkpOUlZaXmJmam5ydnp+QkaKjpKWmp6ipqqusra6voKGys7S1tre4sEoLvLuwvbC6z7G9w7TOz7esxrrMx87EwMHSwNTF2crCyMnW29vN3smo3cKq5NXv4Nfi6e/hyO/s4ez107XV99f427z9/v/w8woMCBBAsaPIgwocKFDBs6fAgxosSJFCtavIgxo8aNPhw7evwIMqTIkSRLmjyJMqXKlSxbunwJM6bMmTRr2ryJM6fOnTx7+vwJNKjQoUSLGj2KNKnSpUybOn0KdU4BADs=',

There is a php program that includes this and does:

$color['C'] = base64_encode("\x00\x0e\x98\x00\xff\xff");
$color['U'] = base64_encode("\x00\x00\xcc\x00\xff\xff");
$color['O'] = base64_encode("\x00\x99\xff\x99\xff\xff");

// Create image instances
     $dest = imagecreate($map_width,$map_height);

   // white background color
     $white = imagecolorallocatealpha($dest, 255, 255, 255, 0);
     imagecolortransparent($dest, $white);

   // mn map on white background
     $src = imagecreatefromstring(base64_decode($mnmap));
     imagecopy($dest, $src, 0, 0, 0, 0, $map_width, $map_height);

// then the counties with changed color.
                 $src = 
imagecreatefromstring(base64_decode(substr($cty,0,16).$color[$ch].substr($cty,24)));
                 imagecopy($dest, $src, 0, 0, 0, 0, $map_width, 
$map_height);

// after all the counties, output the map

     header('Content-Type: image/png');
     imagepng($dest);
     imagedestroy($src);
     imagedestroy($dest);

I looked at the internal structure of the .png files and could not find 
the color.  It is probably in the zipped region of the file. I've 
included a red and a green county so you can see the difference.  It may 
be that I can use the image routines to change the color after the image 
has been created from the string.

Again, Thanks for your help!

On 5/22/2015 1:50 PM, skjorn wrote:
>
> I'm glad you find it helpful. We could add the support for GIF 
> eventually, sure, but it would take some time to handle all these 
> parameters, as you say.
>
> I'm not sure what "string" you're talking about, but the functionality 
> you describe (changing colours) can be achieved using PNG-8 as well. 
> It basically is the same thing -- it has a palette of colours and 
> pixels reference the palette by indices. If you use a standard library 
> (libpng, for example) in your program, you can change the colours in 
> the palette and the whole picture will become different. Assuming you 
> can modify your program.
>
> The only advantage the GIF has over the PNG is animation. And I'm 
> afraid it wouldn't work well with this script. (Well, you could 
> probably export multiple animations with one changed frame.)
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/jwa107/Photoshop-Export-Layers-to-Files-Fast/issues/51#issuecomment-104742251>.
>

-- 
David Cahlander david@cahlander.com Burnsville MN 952-894-5910

**Comment by skjorn** [comment]
I don't get why you encode colours into base64, concatenate and then decode them again into an image that you're using locally, relying on a specific byte offset in the file probably. (I don't know what $cty is and where $ch comes from.)

The easiest thing would be to:

1. Load the file of a county:
imagecreatefromstring()
imagecreatefrompng()
imagecreatefromgif()
whatever...
(That's your $src if I understand correctly.)

2. If needed, change the color table in the $src:
imagecolorset()

3. Copy the county to the destination canvas ($src => $dst):
imagecopy()

That is just very inefficient and hackish code that you have there. Not speaking that the colour is fixed size, so splitting the string and joining it again just to insert the colour is not the best practice overall.
($src = imagecreatefromstring(base64_decode(substr($cty,0,16).$color[$ch].substr($cty,24)));)

But back to the topic. You should definitely be able to rewrite your program to use the indexed PNG and/or GIF for input. Just by glancing at the manual, as I'm not a PHP programmer.

**Comment by cahlander** [comment]
That works just great!  I've changed the program and it now works with 
just the imagecopy().  I had read the description of this program and 
looked at the examples, but until you suggested that it would work, I 
didn't try it.  Some of the confusing logic in what I sent to you was 
due to its use in <canvas> in another program.  I didn't want to read 
all the 87 little files (300bytes each) so I came across the idea of 
storing the files in an include file.  Since the include file needs to 
have data in a readable form, I put it into base64 format.  <img> has a 
format that will take base64 data:

     <img src='data:image/gif;base64,base64_data>

so for this operation I started with base64 data, inserted the color 
change, and used it as the image.

Thanks so much for the help, and  I don't need any GIF conversion.

-- 
David Cahlander david@cahlander.com Burnsville MN 952-894-5910
